File: 892-shortest-subarray-with-sum-at-least-k/shortest-subarray-with-sum-at-least-k.java
class Solution {
    public int shortestSubarray(int[] nums, int k) {
       Deque<Pair<Long, Integer>> d = new ArrayDeque<>();
       long s = 0;
       int result = Integer.MAX_VALUE;
       for (int r = 0; r < nums.length; r++){
            s += nums[r];
            if (s >= k) result = Math.min(result, r + 1);

            while (d.size() > 0 && s - d.getFirst().getKey() >= k){
                Pair<Long, Integer> temp = d.getFirst();
                d.removeFirst();
                result = Math.min(result, r - temp.getValue());
            }

            while (d.size() > 0 && d.getLast().getKey() >= s)
                d.removeLast();

            d.addLast(new Pair<Long, Integer>(s, r));
       }
       return result == Integer.MAX_VALUE ? -1 : result;
    

        //Using Heap
        // long s = 0;
        // int result = Integer.MAX_VALUE;
        // PriorityQueue<Pair<Long, Integer>> pq = new PriorityQueue<>(Comparator.comparing(Pair::getKey));
        // for (int r = 0; r < nums.length; r++){
        //     s += nums[r];
        //     if (s >= k) result = Math.min(result, r + 1);
            
        //     while (pq.size() > 0  && s - pq.peek().getKey() >= k){
        //         Pair<Long, Integer> temp = pq.poll();
        //         result = Math.min(result, r - temp.getValue());
        //     }

        //     pq.add(new Pair<Long, Integer>(s, r));
        // }
        // return result == Integer.MAX_VALUE ? -1 : result;
    }
}

File: 892-shortest-subarray-with-sum-at-least-k/shortest-subarray-with-sum-at-least-k.cpp
class Solution {
public:
    static bool compare(pair<long, int> a, pair<long, int>b){
        return a.first > b.first;
    }
    int shortestSubarray(vector<int>& nums, int k) {
        //Using Heap
        long s = 0;
        int result = INT_MAX;
        priority_queue<pair<long, int>, vector<pair<long, int>>, function<bool(pair<long, int>, pair<long, int>)>> pq(compare);
        for (int r = 0; r < nums.size(); r++){
            s += nums[r];
            if (s >= k) result = min(result, r + 1);
            
            while (pq.size() > 0  && s - pq.top().first >= k){
                pair<long, int> temp = pq.top();
                pq.pop();
                result = min(result, r - temp.second);
            }

            pq.push(pair<long, int>{s, r});
        }
        return result == INT_MAX ? -1 : result;
    }
};